[fx] Add a pass to fetch attributes of nn.Module to fx.node

842974287 opened the pull request:
Summary: Fetch the parameters that are needed for lowering from nn.Module to fx.node for leaf_modules.
A `module_fetch_book` maintains the names of attribute that we want to fetch for a specific nn.Module. `_version` is used to ensure that the fetch book is aligned with the module.

Test Plan: A test `test_fetch` is added to test_fx_experimental.py.

Differential Revision: D24957142

jamesr66a commented on torch/fx/experimental/param_fetch.py:
```diff
@@ -0,0 +1,42 @@
+The first integer in the tuple is the version number of the nn.Module class when we create the parameter list.
+If there's a version mismatch then it means the parameter names in the book might be mismatched with nn.Module.
+"""
+module_fetch_book: Dict[str, Tuple[int, List[str]]] = {
```
You should key on the `Type`s themselves, not their string `__name__`:

```
from typing import Type

<...>

module_fetch_book: Dict[Type, Tuple[int, List[str]]] = {
    torch.nn.Linear: (1, ["weight", "bias"]),
    torch.nn.Conv2d: (1, ["weight", "bias", "kernel_size", "stride", "padding", "dilation", "groups", "padding_mode"]),
    torch.nn.BatchNorm2d: (2, ["weight", "bias", "running_mean", "running_var"]),
}
```

Using `__name__` isn't safe, because it's only the last atom of the qualified name. Alternatively we could match on `torch.typename(m)`, but it's probably better to not project into string space at all and just match directly on types

jamesr66a commented on torch/fx/experimental/param_fetch.py:
```diff
@@ -0,0 +1,42 @@
+    after checking module's version is compatible with the `module_fetch_book`.
+    """
+    parameters: Dict[str, Any] = {}
+    parameters["name"] = type(mod).__name__
```
```
    parameters["name"] = torch.typename(mod)
```

To get the fully-qualified name

jamesr66a commented on torch/fx/experimental/param_fetch.py:
```diff
@@ -0,0 +1,42 @@
+    parameters: Dict[str, Any] = {}
+    parameters["name"] = type(mod).__name__
+
+    if type(mod).__name__ in module_fetch_book:
```
With the change above, this will become

```
    if type(mod) in module_fetch_book:
        _version, param_to_fetch = module_fetch_book[type(mod)]

```

jamesr66a commented on torch/fx/experimental/param_fetch.py:
```diff
@@ -0,0 +1,42 @@
+    parameters["name"] = type(mod).__name__
+
+    if type(mod).__name__ in module_fetch_book:
+        _version, param_to_fetch = module_fetch_book[type(mod).__name__]
```
Convention is leading underscore indicates a private name, but since this is a local variable that doesn't really make sense, so let's just name this `version`

jamesr66a commented on torch/fx/experimental/param_fetch.py:
```diff
@@ -0,0 +1,42 @@
+    if type(mod).__name__ in module_fetch_book:
+        _version, param_to_fetch = module_fetch_book[type(mod).__name__]
+        if _version != mod._version:
+            print(f"Fetcher version {_version} try to fetch {type(mod).__name__} version {mod._version}")
```
Is this an error condition? If so,

```
            raise RuntimeError(f"Fetcher version {_version} try to fetch {type(mod).__name__} version {mod._version}")
```

If not, use `warnings.warn`.

jamesr66a replied:
We should also be checking that the `Module` we're processing has `version` at *or below* the one we support. Right now, with the 3 modules supported (Linear, Conv2d, and BatchNorm2d) it doesn't matter because the attributes of interest are all the same, but in the future we may need `module_fetch_book` to also contain functions that normalize the retrieved attributes across older versions

842974287 replied:
>contain functions that normalize the retrieved attributes across older versions

Sorry I'm a little bit confused. Do you mean when we processing `Module` with older versions, the function would come out and match the names of attribute in current version to the older version?

jamesr66a replied:
Yes, that's correct

842974287 replied:
I've added a function to each entry. Right now for these three modules the function doesn't actually do anything.

jamesr66a commented on torch/fx/experimental/param_fetch.py:
```diff
@@ -0,0 +1,42 @@
+    """If `mod` is in `module_fetch_book`, fetch the mod's attributes that in the `module_fetch_book`
+    after checking module's version is compatible with the `module_fetch_book`.
+    """
+    parameters: Dict[str, Any] = {}
```
The term "parameter" has a very specific meaning in PyTorch (trainable tensors), so please call this something else like `attrs`

jamesr66a commented on torch/fx/experimental/param_fetch.py:
```diff
@@ -0,0 +1,42 @@
+    "BatchNorm2d": (2, ["weight", "bias", "running_mean", "running_var"]),
+}
+
+def fetch_module(mod: nn.Module) -> Dict[str, Any]:
```
"Fetch module" implies you're going to return a `Module` instance, but you're not. This is probably better named something like `extract_attrs_for_lowering`

jamesr66a commented on torch/fx/experimental/param_fetch.py:
```diff
@@ -0,0 +1,42 @@
+
+    return parameters
+
+def fetch(fx_module: GraphModule):
```
Similarly, this could be named better. Something like `lift_lowering_parameters_to_nodes`

jamesr66a reviewed (commented):
Cool!

jamesr66a commented on test/test_fx_experimental.py:
```diff
@@ -550,6 +551,28 @@ def foo(x):
+        for node in traced.graph.nodes:
+            if node.op == "call_module":
+                assert hasattr(node, "attrs_for_lowering")
+                _, para_list = module_fetch_book[locate(node.attrs_for_lowering["name"])]
```
As part of good testing practice, you shouldn't use the code you're trying to test as a reference in the test. I suggest just having a list of expected attributes for the Conv and BatchNorm node and assert that they are present manually.

jamesr66a commented on torch/fx/node.py:
```diff
@@ -53,6 +53,9 @@ def __init__(self, graph: 'Graph', name: str, op: str, target: Target,
         self.type : Optional[Any] = type
+        # Attrbutes that are needed to lowering the fx node.
+        # This dict is filled by param_fetch.
+        self.attrs_for_lowering : Dict[str, Any] = {}
```
Please don't add this code to FX core, it is not generally applicable. You can either monkey-patch this attribute, or create a subclass of `Node` to inject into your `Graph`

842974287 replied:
Ah I see, I'll try monkey-patch first (hopefully no mypy error

jamesr66a commented on test/test_fx_experimental.py:
```diff
@@ -595,6 +596,39 @@ def foo(x):
+        for node in traced.graph.nodes:
+            if node.op == "call_module":
+                assert hasattr(node, "attrs_for_lowering")
+                para_list = attrs_for_lowering[locate(node.attrs_for_lowering["name"])]
```
How about we just use string keys in `attrs_for_lowering` in this test instead of using `locate`?

842974287 replied:
Sounds good!

jamesr66a commented on torch/fx/experimental/graph_manipulation.py:
```diff
@@ -122,19 +123,17 @@ def serialize_weight(tensor: torch.Tensor) -> Dict:
-        for constant in mod.__constants__:
-            parameters[constant] = str(getattr(mod, constant))
+
+    for p_name, p_value in node.attrs_for_lowering.items():  # type: ignore
```
Why is there a `# type: ignore` here?

842974287 replied:
Oh this is because I'm doing monkey-patch to add `attrs_for_lowering` which would cause the mypy test fail with error `node doesn't has attribute attrs_for_lowering`.

jamesr66a commented on torch/fx/experimental/param_fetch.py:
```diff
@@ -0,0 +1,50 @@
+    if type(mod) in module_fetch_book:
+        version, param_to_fetch, matching_method = module_fetch_book[type(mod)]
+        if version < mod._version:
+            raise RuntimeError(f"Fetcher version {version} try to fetch {torch.typename(mod)} version {mod._version}")
```
This error message should also include instructions for what to do to proceed. In this case, likely the user should report a bug to the accelerator team

842974287 replied:
Sounds good!

jamesr66a reviewed (commented):
LGTM! @gcatron @jfix71 do you want to take a look?

jfix71 commented on torch/fx/experimental/param_fetch.py:
```diff
@@ -0,0 +1,50 @@
+    attrs_for_lowering: Dict[str, Any] = {}
+    attrs_for_lowering["name"] = torch.typename(mod)
+
+    if type(mod) in module_fetch_book:
```
Should we raise an error if we don't handle the type?

842974287 replied:
Sounds good!

jfix71 reviewed (commented):
Cool! Mostly LGTM, one comment

jackm321 commented on torch/fx/experimental/param_fetch.py:
```diff
@@ -0,0 +1,59 @@
+        version, param_to_fetch, matching_method = module_fetch_book[type(mod)]
+        if version < mod._version:
+            raise RuntimeError(f"Fetcher version {version} try to fetch {torch.typename(mod)} version {mod._version}, "
+                               "please upgrade the module_fetch_book or report a bug to AIACC team.")
```
folks in OSS won't know who AIACC team is. Should we just ask for a gh issue or something?

842974287 replied:
Sounds good! I added another option as "open an issue and @842974287".

jfix71 commented on torch/fx/experimental/param_fetch.py:
```diff
@@ -0,0 +1,60 @@
+        version, param_to_fetch, matching_method = module_fetch_book[type(mod)]
+        if version < mod._version:
+            raise RuntimeError(f"Fetcher version {version} try to fetch {torch.typename(mod)} version {mod._version}, "
+                               "please upgrade the module_fetch_book, open an issue and @842974287"
+                               "or report a bug to AIACC team directly.")
+        for attr in param_to_fetch:
+            attrs_for_lowering[attr] = getattr(mod, matching_method(attr, mod._version))
+    else:
+        raise RuntimeError(f"{torch.typename(mod)} is not in the module_fetch_book yet, "
+                           "please add it to the module_fetch_book, open an issue and @842974287"
+                           "or report a bug to AIACC team directly.")
+    return attrs_for_lowering
```
nit: You didn't add a space between `@842974287` and the next line `or`, so it'll appear as `@842974287or` 🙂

842974287 replied:
Nice catch!

jfix71 reviewed (approved):
🚀

842974287 committed:
Add a pass to fetch attributes of nn.Module to fx.node (#47935)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/47935

Fetch the parameters that are needed for lowering from nn.Module to fx.node for leaf_modules.

Test Plan: A test `test_fetch` is added to test_fx_experimental.py.

Reviewed By: jfix71

Differential Revision: D24957142

fbshipit-source-id: c6eadd79594763dcec67854cf8753ab41f1d2f1a
